protocol/chat_persistence_unviewed_counts.go

```diff
@@ -0,0 +1,136 @@
+package protocol
+
+import (
+	"context"
+	"database/sql"
+	"fmt"
+	"strings"
+)
+
+var msgNotHidden = "NOT(m1.hide) AND NOT(m1.deleted) AND NOT(m1.deleted_for_me)"
+
+func (db sqlitePersistence) executeTx(tx *sql.Tx, fn func(tx *sql.Tx) error) (err error) {
+	if tx == nil {
+		tx, err = db.db.BeginTx(context.Background(), &sql.TxOptions{})
+		if err != nil {
+			return err
+		}
+		defer func() {
+			if err != nil {
+				// don't shadow original error
+				_ = tx.Rollback()
+			} else {
+				err = tx.Commit()
+			}
+		}()
+	}
+
+	return fn(tx)
+}
+
+func (db sqlitePersistence) updateAllChatsUnviewedCounts(tx *sql.Tx) error {
+	return db.executeTx(tx, func(tx *sql.Tx) error {
+		// Recalculate denormalized fields
+		_, err := tx.Exec(
+			fmt.Sprintf(`
+		UPDATE chats
+		SET
+			unviewed_message_count =
+				(SELECT COUNT(1)
+				FROM user_messages m1
+				WHERE local_chat_id = chats.id
+				AND NOT(seen) AND %s),
+			unviewed_mentions_count =
+				(SELECT COUNT(1)
+				FROM user_messages m1
+				WHERE local_chat_id = chats.id
+				AND NOT(seen) AND %s AND (mentioned OR replied)),
+			first_unviewed_message_id =
+				(SELECT id
+				FROM user_messages m1
+				WHERE local_chat_id = chats.id
+				AND NOT(seen) AND %s
+				ORDER BY clock_value ASC
+				LIMIT 1),
+			highlight = 0`, msgNotHidden, msgNotHidden, msgNotHidden))
+
+		return err
+	})
+}
+
+func (db sqlitePersistence) updateChatUnviewedCounts(chatID string, tx *sql.Tx) error {
+	return db.executeTx(tx, func(tx *sql.Tx) error {
+		// Recalculate denormalized fields
+		_, err := tx.Exec(
+			fmt.Sprintf(`
+		UPDATE chats
+		SET
+			unviewed_message_count =
+				(SELECT COUNT(1)
+				FROM user_messages m1
+				WHERE local_chat_id = ?
+				AND NOT(seen) AND %s),
+			unviewed_mentions_count =
+				(SELECT COUNT(1)
+				FROM user_messages m1
+				WHERE local_chat_id = ?
+				AND NOT(seen) AND %s AND (mentioned OR replied)),
+			first_unviewed_message_id =
+				(SELECT id
+				FROM user_messages m1
+				WHERE local_chat_id = ?
+				AND NOT(seen) AND %s
+				ORDER BY clock_value ASC
+				LIMIT 1),
+			highlight = 0
```

Comment:
The top of the query is the same as above I think, so maybe you can extract it to a constant or function